arroyo/processing/strategies/filter.py

```diff
@@ -28,8 +39,30 @@ def poll(self) -> None:
     def submit(self, message: Message[TPayload]) -> None:
         assert not self.__closed
 
+        now = time.time()
+
         if self.__test_function(message):
+            if self.__commit_policy_state is not None:
+                self.__flush_uncommitted_offsets(now)
             self.__next_step.submit(message)
```

Comment:
>Since there are enough messages actually passing the test for committing to still regularly happen

But this PR is about scenarios where this isn't the case, right? You don't have to pass commitpolicy into FilterStep. If you don't pass one, FilterStep works exactly like it does on master, and that version of it is already pretty much perfect for that scenario.

----

I see three paths forward:

1. we expect that all partitions regularly receive messages that pass the filter. However, in that case one could just not pass a commit policy into FilterStep and the code works exactly as in master, which already works fine under such conditions.
2. we make the flushing conditional based on the given commit policy (just like in the codepath on line 51). this will cause offsets to be committed out of order, i.e. offset 2 can be committed after offset 3. we can make further changes in the processor or elsewhere to sanitize those commits
3. we attach the to-be-flushed committable offsets to the message that we submit immediately after this one (on line 47). This requires us to change the concrete type of `message.value` which strategies could rely on (even though they probably shouldn't)

Ideally the filterstep would be able to exhibit optimal performance regardless of whether the majority of messages are filtered, I just couldn't think of a way to make that happen.

Reply:
How about we remove any offsets staged in `self.__uncommitted_offsets` for that partition if the `test_function` passes and we have a message to actually send through the pipeline. That way, we can safely throttle flushing the uncommitted offsets so it actually respects the commit policy and we won't get offsets committed out of order. The rate of filtered messages being forwarded would be the same regardless of how many messages in between actually do pass the filtering.